iroh-net/src/hp/netcheck/reportgen/probes.rs

```diff
@@ -0,0 +1,724 @@
+pub(super) enum Probe {
+    #[display("Ipv4 after {delay:?} to {node}")]
+    StunIpv4 {
+        /// When the probe is started, relative to the time that `get_report` is called.
+        /// One probe in each `ProbePlan` should have a delay of 0. Non-zero values
+        /// are for retries on UDP loss or timeout.
+        delay: Duration,
+
+        /// The name of the node name. DERP node names are globally
+        /// unique so there's no region ID.
+        node: Arc<DerpNode>,
+    },
+    #[display("Ipv6 after {delay:?} to {node}")]
+    StunIpv6 {
+        delay: Duration,
+        node: Arc<DerpNode>,
+    },
+    // TODO: Probably can remove DerpRegion since the DerpNode already contains the region
+    // ID which can then be looked up in the DerpMap.  But Https isn't even implemented
+    // right now so leave it.
+    #[display("Https after {delay:?} to {node}")]
+    Https {
+        delay: Duration,
+        node: Arc<DerpNode>,
+        region: DerpRegion,
+    },
+    #[display("Icmp after {delay:?} to {node}")]
+    Icmp {
+        delay: Duration,
+        node: Arc<DerpNode>,
+    },
+}
+
+impl Probe {
+    pub(super) fn delay(&self) -> Duration {
+        match self {
+            Probe::StunIpv4 { delay, .. }
+            | Probe::StunIpv6 { delay, .. }
+            | Probe::Https { delay, .. }
+            | Probe::Icmp { delay, .. } => *delay,
+        }
+    }
+
+    pub(super) fn proto(&self) -> ProbeProto {
+        match self {
+            Probe::StunIpv4 { .. } => ProbeProto::StunIpv4,
+            Probe::StunIpv6 { .. } => ProbeProto::StunIpv6,
+            Probe::Https { .. } => ProbeProto::Https,
+            Probe::Icmp { .. } => ProbeProto::Icmp,
+        }
+    }
+
+    pub(super) fn node(&self) -> &Arc<DerpNode> {
+        match self {
+            Probe::StunIpv4 { node, .. }
+            | Probe::StunIpv6 { node, .. }
+            | Probe::Https { node, .. }
+            | Probe::Icmp { node, .. } => node,
+        }
+    }
+}
+
+/// A probe set is a sequence of similar [`Probe`]s with delays between them.
+///
+/// The probes are to the same region and of the same [`ProbeProto`] but will have different
+/// delays.  The delays are effectively retries, though they do not wait for the previous
+/// probe to be finished.  The first successful probe will cancel all other probes in the
+/// set.
+///
+/// This is a lot of type-safety by convention.  It would be so much nicer to have this
+/// compile-time checked but that introduces a giant mess of generics and traits and
+/// associated exploding types.
+///
+/// A [`ProbeSet`] implements [`IntoIterator`] similar to how [`Vec`] does.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub(super) struct ProbeSet {
+    /// Name of this probe set, informational without uniqueness guarantee.
+    name: String,
+    /// The [`ProbeProto`] all the probes in this set have.
+    proto: ProbeProto,
```

Comment:
I still hate this name..(not your fault)

Reply:
heh, I did get used to it a little.